### PR TITLE
[ML] Add information about samples per node to the tree

### LINKED ISSUE #850: [ML] Add information about samples per node to the tree
I propose to extend the node definition by adding information about the number of training samples that passed through the node. This information provide one of the simpler feature importance indicators (see[ xgboost cover](https://xgboost.readthedocs.io/en/latest/R-package/discoverYourData.html#measure-feature-importance)). It also helps to test and  prevent degenerated trees such as those described in #849 .

#### Sub-Tasks
- [x] Update schema definition
- [x] Update `Node` logic
- [x] Update inference model code
- [x] Update persist/restore schema version
- [x] Add enhancement note

### PR DESCRIPTION
**Pull request by valeriy42**
This PR  extends the definition of  the tree node by adding information about the number of training samples that passed through the node (`numberSamples` or `number_samples`). The json schema for inference model is adjusted accordingly.

Since this change the schema for persist/restore of the tree implementation, I bumped the version and removed 7.5 and 7.6 from the list of supported version. My reasoning: restoring from old schema and setting number samples to 0 would break feature importance at inference time.

I also adjust feature importance computation to use pre-computed number samples instead of recomputing it on the fly.

Closes #850

### CODE REVIEW

**Comment by valeriy42** [comment]
retest

File: lib/maths/CBoostedTreeLeafNodeStatistics.cc [hunk 1]
```diff
@@ -398,5 +404,9 @@ CBoostedTreeLeafNodeStatistics::computeBestSplitStatistics(const TRegularization
 
     return result;
 }
+
+std::size_t CBoostedTreeLeafNodeStatistics::numberSamples() const {
+    return static_cast<std::size_t>(m_RowMask.manhattan());
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeLeafNodeStatistics.cc]: Note if you move to a standalone step to write sample counts to the tree as below, then this obviously no longer applies.

File: include/maths/CBoostedTree.h [hunk 2]
```diff
@@ -265,8 +265,9 @@ class MATHS_EXPORT CBoostedTreeNode final {
                              bool assignMissingToLeft,
                              double nodeValue,
                              double gain,
-                             TOptionalNodeIndex leftChild,
-                             TOptionalNodeIndex rightChild) = 0;
+                             std::size_t numberSamples,
+                             maths::CBoostedTreeNode::TOptionalNodeIndex leftChild,
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply include/maths/CBoostedTree.h]: fixed

File: include/maths/CBoostedTree.h [hunk 3]
```diff
@@ -265,8 +265,9 @@ class MATHS_EXPORT CBoostedTreeNode final {
...
+                             maths::CBoostedTreeNode::TOptionalNodeIndex rightChild) = 0;
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply include/maths/CBoostedTree.h]: fixed

File: include/maths/CBoostedTree.h [hunk 4]
```diff
@@ -306,19 +307,26 @@ class MATHS_EXPORT CBoostedTreeNode final {
     //! Get the total curvature at the rows below this node.
     double curvature() const { return m_Curvature; }
 
+    //! Set the number of samples to \p value.
+    void numberSamples(size_t value);
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply include/maths/CBoostedTree.h]: fixed

File: include/maths/CBoostedTree.h [hunk 5]
```diff
@@ -306,19 +307,26 @@ class MATHS_EXPORT CBoostedTreeNode final {
...
+
+    //! Get number of samples affected by the node.
+    size_t numberSamples() const;
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply include/maths/CBoostedTree.h]: fixed

File: include/maths/CBoostedTree.h [hunk 6]
```diff
@@ -306,19 +307,26 @@ class MATHS_EXPORT CBoostedTreeNode final {
...
     TNodeIndex rightChildIndex() const { return m_RightChild.get(); }
 
     //! Split this node and add its child nodes to \p tree.
-    TSizeSizePr split(std::size_t splitFeature,
-                      double splitValue,
-                      bool assignMissingToLeft,
-                      double gain,
-                      double curvature,
-                      TNodeVec& tree);
+    CBoostedTreeNode::TSizeSizePr split(std::size_t splitFeature,
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply include/maths/CBoostedTree.h]: fixed

File: include/maths/CTreeShapFeatureImportance.h [hunk 7]
```diff
@@ -40,17 +40,14 @@ class MATHS_EXPORT CTreeShapFeatureImportance {
     //! by \p offset.
     void shap(core::CDataFrame& frame, const CDataFrameCategoryEncoder& encoder, std::size_t offset);
 
-    //! Compute number of training samples from \p frame that pass every node in the \p tree.
-    static TDoubleVec samplesPerNode(const TTree& tree,
-                                     const core::CDataFrame& frame,
-                                     const CDataFrameCategoryEncoder& encoder,
-                                     std::size_t numThreads);
+    //! Return the array with number of training samples affected by every node of the \p tree.
+    static CTreeShapFeatureImportance::TDoubleVec numberSamples(const TTree& tree);
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply include/maths/CTreeShapFeatureImportance.h]: fixed

File: lib/maths/CBoostedTreeImpl.cc [hunk 8]
```diff
@@ -1274,15 +1276,7 @@ void CBoostedTreeImpl::acceptPersistInserter(core::CStatePersistInserter& insert
...
 bool CBoostedTreeImpl::acceptRestoreTraverser(core::CStateRestoreTraverser& traverser) {
-    if (traverser.name() == VERSION_7_5_TAG) {
-        // Force downsample factor to 1.0.
-        m_DownsampleFactorOverride = 1.0;
-        m_DownsampleFactor = 1.0;
-        m_BestHyperparameters.downsampleFactor(1.0);
-        // We can't stop cross-validation early because we haven't gathered the
-        // per fold test losses.
-        m_StopCrossValidationEarly = false;
-    } else if (traverser.name() != VERSION_7_6_TAG) {
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply lib/maths/CBoostedTreeImpl.cc]: :fire:

File: lib/maths/CBoostedTreeLeafNodeStatistics.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (valeriy42) [reply lib/maths/CBoostedTreeLeafNodeStatistics.cc]: I rewrote the part to have a standalone method after training the final tree.

File: lib/maths/CTreeShapFeatureImportance.cc [hunk 9]
```diff
@@ -37,72 +36,44 @@ void CTreeShapFeatureImportance::shap(core::CDataFrame& frame,
...
-    TDoubleVec totalSamplesPerNode{std::move(state[0].s_FunctionState)};
-    for (std::size_t i = 1; i < state.size(); ++i) {
-        for (std::size_t nodeIndex = 0; nodeIndex < totalSamplesPerNode.size(); ++nodeIndex) {
-            totalSamplesPerNode[nodeIndex] += state[i].s_FunctionState[nodeIndex];
-        }
+CTreeShapFeatureImportance::numberSamples(const TTree& tree) {
+    TDoubleVec numberSamples;
+    numberSamples.reserve(tree.size());
+    for (std::size_t i = 0; i < tree.size(); ++i) {
+        numberSamples.emplace_back(static_cast<double>(tree[i].numberSamples()));
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply lib/maths/CTreeShapFeatureImportance.cc]: I did as you suggested :+1:

File: lib/api/unittest/CDataFrameAnalyzerFeatureImportanceTest.cc [hunk 10]
```diff
@@ -299,8 +299,9 @@ BOOST_FIXTURE_TEST_CASE(testRegressionFeatureImportanceNoImportance, SFixture) {
             // c1 explains 95% of the prediction value, i.e. the difference from the prediction is less than 2%.
             BOOST_REQUIRE_CLOSE(c1, prediction, 5.0);
             for (const auto& feature : {"c2", "c3", "c4"}) {
-                BOOST_REQUIRE_SMALL(readShapValue(result, feature), 2.0);
-                cNoImportanceMean.add(std::fabs(readShapValue(result, feature)));
+                double c = readShapValue(result, feature);
+                BOOST_REQUIRE_SMALL(c, 2.5);
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply lib/api/unittest/CDataFrameAnalyzerFeatureImportanceTest.cc]: I changed the threshold back.

File: lib/maths/CBoostedTreeImpl.cc [hunk 11]
```diff
@@ -721,7 +766,7 @@ CBoostedTreeImpl::trainTree(core::CDataFrame& frame,
         std::tie(leftChild, rightChild) =
             leaf->split(leftChildId, rightChildId, m_NumberThreads, frame, *m_Encoder,
                         m_Regularization, candidateSplits, this->featureBag(),
-                        tree[leaf->id()], leftChildHasFewerRows);
+                        tree[leaf->id()], leftChildHasFewerRows, tree);
```
Comment (tveasey) [inline lib/maths/CBoostedTreeImpl.cc]: No longer needed
```suggestion
                        tree[leaf->id()], leftChildHasFewerRows);
```

File: lib/maths/CBoostedTreeLeafNodeStatistics.cc [hunk 12]
```diff
@@ -134,7 +134,8 @@ CBoostedTreeLeafNodeStatistics::split(std::size_t leftChildId,
                                       const TImmutableRadixSetVec& candidateSplits,
                                       const TSizeVec& featureBag,
                                       const CBoostedTreeNode& split,
-                                      bool leftChildHasFewerRows) {
+                                      bool leftChildHasFewerRows,
+                                      TNodeVec& tree) {
```
Comment (tveasey) [inline lib/maths/CBoostedTreeLeafNodeStatistics.cc]: All the changes in this file should be reverted now.

File: lib/maths/CTreeShapFeatureImportance.cc [hunk 13]
```diff
@@ -170,14 +141,14 @@ void CTreeShapFeatureImportance::shapRecursive(const TTree& tree,
             CTreeShapFeatureImportance::unwindPath(splitPath, pathIndex);
         }
 
-        double hotFractionZero = samplesPerNode[hotIndex] / samplesPerNode[nodeIndex];
-        double coldFractionZero = samplesPerNode[coldIndex] / samplesPerNode[nodeIndex];
+        double hotFractionZero = numberSamples[hotIndex] / numberSamples[nodeIndex];
```
Comment (tveasey) [inline lib/maths/CTreeShapFeatureImportance.cc]: Surely just (etc)
```suggestion
        double hotFractionZero{static_cast<double>(tree[hotIndex].numberSamples()) /
                               static_cast<double>(tree[nodeIndex].numberSamples())};
```

File: include/maths/CTreeShapFeatureImportance.h [hunk 14]
```diff
@@ -126,7 +123,7 @@ class MATHS_EXPORT CTreeShapFeatureImportance {
     //! Recursively traverses all pathes in the \p tree and updated SHAP values once it hits a leaf.
     //! Ref. Algorithm 2 in the paper by Lundberg et al.
     void shapRecursive(const TTree& tree,
-                       const TDoubleVec& samplesPerNode,
+                       const TDoubleVec& numberSamples,
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply include/maths/CTreeShapFeatureImportance.h]: You are correct, I changed it in 21e714dd

File: lib/maths/CTreeShapFeatureImportance.cc (hunk 13 above)
Comment (valeriy42) [reply lib/maths/CTreeShapFeatureImportance.cc]: Changed in 21e714dd

File: lib/maths/CBoostedTreeImpl.cc [hunk 15]
```diff
@@ -258,6 +262,47 @@ void CBoostedTreeImpl::train(core::CDataFrame& frame,
...
+        if (tree.size() == 1) {
+            tree[0].numberSamples(frame.numberRows());
+        } else {
+            auto result = frame.readRows(
+                m_NumberThreads,
+                core::bindRetrievableState(
+                    [&](TSizeVec& samplesPerNode, const TRowItr& beginRows, const TRowItr& endRows) {
+                        for (auto row = beginRows; row != endRows; ++row) {
+                            auto encodedRow{m_Encoder->encode(*row)};
+                            const CBoostedTreeNode* node{&tree[0]};
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply lib/maths/CBoostedTreeImpl.cc]: Fixed in 3fbb75e5

File: lib/maths/CBoostedTreeImpl.cc (hunk 11 above)
Comment (valeriy42) [reply lib/maths/CBoostedTreeImpl.cc]: Fixed in 3fbb75e5

File: lib/maths/CBoostedTreeLeafNodeStatistics.cc (hunk 12 above)
Comment (valeriy42) [reply lib/maths/CBoostedTreeLeafNodeStatistics.cc]: Reverted in c30908e0

File: lib/maths/CBoostedTreeImpl.cc [hunk 16]
```diff
@@ -232,6 +232,10 @@ void CBoostedTreeImpl::train(core::CDataFrame& frame,
         std::tie(m_BestForest, std::ignore) =
             this->trainForest(frame, allTrainingRowsMask, allTrainingRowsMask,
                               m_TrainingProgress, recordMemoryUsage);
+
+        // populate numberSamples field in the final forest
+        this->computeNumberSamples(frame);
+
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply lib/maths/CBoostedTreeImpl.cc]: Moved in 93a4432f

File: include/maths/CTreeShapFeatureImportance.h [hunk 17]
```diff
@@ -146,7 +143,7 @@ class MATHS_EXPORT CTreeShapFeatureImportance {
 private:
     TTreeVec m_Trees;
     std::size_t m_NumberThreads;
-    TDoubleVecVec m_SamplesPerNode;
+    TDoubleVecVec m_NumberSamples;
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply include/maths/CTreeShapFeatureImportance.h]: Removed in 21e714dd

**Comment by valeriy42** [comment]
Thank you, @tveasey for the review. I addressed your comments and removed unnecessary code. Let me know if everything is ok now.

**Comment by valeriy42** [comment]
retest

**Pull request merged**